core/domain/exp_domain.py

```diff
@@ -475,14 +475,19 @@ def __init__(self, change_list):
 
 
 class StateVersionSpan(python_utils.OBJECT):
-    """A consecutive span of versions in which a state remains "equivalent".
+    """A consecutive span of exploration versions with equivalent state content.
 
-    Equivalence between spans are defined using a user-defined predicate, passed
-    in to the constructor. By default, all states are equivalent to each other.
+    Equivalence between state versions are determined by the user-defined
+    predicate passed in to the constructor. By default, all versions of a state
+    are equal.
 
-    Spans must be extended in increasing version-order. It is an error to extend
-    a span with a version older than what it already contains; exceptions are
-    raised to enforce this contract.
+    Spans can only be extended using the extend_or_split() method. It enforces
+    the invariant: all versions of the state in a span are equivalent. Spans
+    must be extended in increasing and consecutive version-order.
+
+    Equivalence has a prerequisite: the names of the state must match across all
```

Comment:
I was trying to communicate that the predicate must pass _and_ the name of the states must match.

Reply:
I'm still confused, sorry. I sort of get what you mean, but when you say "an ExplorationVersionsDiff must be used to resolve the discrepancy", it raises the question of "what happens if that doesn't happen?"

Reply:
The `extend_or_split` method requires it as a positional argument, so there's no way to "not use it".

Reply:
Right -- but the phrasing of your docstring suggests that there are also possible (wrong) approaches that involve not using the ExplorationVersionsDiff.

Here's an attempt at rewriting the paragraph, feel free to tweak it as you see fit:

"State identity is determined based on the name of the state, in the context of the exploration version provided on initialization of this class. Since the state name may change across versions, any methods that extend the span (such as extend_or_split()) should require an exp_version_diff parameter in order to track such changes."

Reply:
Hmm, I've re-ordered the paragraphs to emphasize the importance of `extend_or_split`, and added more qualifiers to which `exp_version_diff` is being referenced. PTAL!